source/material_model/averaging.cc

```diff
@@ -0,0 +1,412 @@
+            {
+              Assert (values_out[i] >= 0,
+                      ExcMessage ("Computing the geometric average "
+                                  "only makes sense for non-negative "
+                                  "quantities."));
+              prod *= values_out[i];
+            }
+
+          const double average = std::pow (prod, 1./N);
+          for (unsigned int i=0; i<N; ++i)
+            values_out[i] = average;
+        }
+      else if (averaging_operation_lower ==  "pick largest")
+        {
+          double max = -std::numeric_limits<double>::max();
+          for (unsigned int i=0; i<N; ++i)
+            max = std::max(max, values_out[i]);
+
+          for (unsigned int i=0; i<N; ++i)
+            values_out[i] = max;
+          //break;
+        }
+      else if (averaging_operation_lower == "nwd arithmetic average")
+        {
+          double sum_value = 0;
+          double sum_weights = 0;
+          std::vector<double> temp_values(N,0);
+          double max_distance = 0;
+          for (unsigned int i=0; i<N; ++i)
+            {
+              sum_value = 0;
+              sum_weights = 0;
+              double h = bell_shape_limit;
+              double alfad = 0;
+              if (dim==2)
+                {
+                  alfad = 5/(numbers::PI * h * h);
+                }
+              else
+                {
+                  alfad = 106/(16 * numbers::PI * h * h *h);
+                }
+
+              for (unsigned int j=0; j<N; ++j)
+                {
+                  if (position[i].distance(position[j]) > max_distance)
+                    max_distance = position[i].distance(position[j]);
+                }
+              for (unsigned int j=0; j<N; ++j)
+                {
+                  double distance = position[i].distance(position[j])/max_distance;
+                  double weight = alfad * ((1 + 3 * (distance / h)) * (1 - (distance / h)) * (1 - (distance / h)) * (1 - (distance / h)));
+                  if (distance / h > 1)
+                    weight = 0;
+                  sum_value += weight * values_out[j];
+                  sum_weights += weight;
+                }
+
+              const double average = sum_value / sum_weights;
+
+              temp_values[i] = average;
+            };
+          for (unsigned int i = 0; i<N; ++i)
+            {
+              values_out[i] = temp_values[i];
+            }
+        }
+      else if (averaging_operation_lower == "nwd harmonic average")
+        {
+          double sum_value = 0;
+          double sum_weights = 0;
+          std::vector<double> temp_values(N,0);
+          double max_distance = 0;
+          for (unsigned int i=0; i<N; ++i)
+            {
+              sum_value = 0;
+              sum_weights = 0;
+              double h = bell_shape_limit;
+              double alfad = 0;
+              if (dim==2)
+                {
+                  alfad = 5/(numbers::PI * h * h);
```

Comment:
since you have this variable in every block and it has the same value every time, just move it up and initialize it only once.